common/unsafe/src/main/java/org/apache/spark/unsafe/memory/HeapMemoryAllocator.java

```diff
@@ -47,23 +47,29 @@ private boolean shouldPool(long size) {
 
   @Override
   public MemoryBlock allocate(long size) throws OutOfMemoryError {
-    if (shouldPool(size)) {
+    int arraySize = (int)((size + 7) / 8);
+    long alignedSize = arraySize * 8;
+    if (shouldPool(alignedSize)) {
       synchronized (this) {
-        final LinkedList<WeakReference<MemoryBlock>> pool = bufferPoolsBySize.get(size);
+        final LinkedList<WeakReference<MemoryBlock>> pool = bufferPoolsBySize.get(alignedSize);
         if (pool != null) {
           while (!pool.isEmpty()) {
             final WeakReference<MemoryBlock> blockReference = pool.pop();
             final MemoryBlock memory = blockReference.get();
             if (memory != null) {
-              assert (memory.size() == size);
+              assert ((int)((memory.size() + 7) / 8) == arraySize);
+              memory.resetSize(size);
```

Comment:
Hmm, from my understanding the size of `MemoryBlock` is always the actual size, not the aligned size, so looks like we dont need to reset the size here.

Reply:
yes, `MemoryBlock` is always the actual size, if we reuse the previous memory,we should modify the  size of `MemoryBlock`

Reply:
I got it, thanks for the explanation.